rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+time is being spent, with specific emphasis on off-CPU operations**.
+
+For a concrete example, consider the file source. It has a run loop that
+consists at a high level of discovering new files, reading from tracked files,
+sending the output upstream, and sleeping if all files were idle. This implies
+a good starting point for instrumenting it with spans would be to add an overall
+loop iteration span, and child spans for each of those four primary stages. Even
+with only those in place, the tracing data is immediately more useful than
+either metrics or profiling for a number of real world issues we've encountered.
+For stages with multiple interesting sub-stages (e.g. collecting paths vs
+fingerprinting during file discovery), one additional layer of child spans can
+be added to further increase the amount of information available for debugging.
+
+In addition to the question of what spans to add, we need to determine the log
+level at which these spans should operate. Because we've been using `tracing`
+primarily as a traditional logging library to this point, leveling spans and
+having those levels coupled to logging statements can feel a bit weird. To keep
+things relatively simple, **we propose that the spans discussed here be given
+a default level of `INFO`**. This will make them available without the
+additional noise that comes with increased logging levels, and should not cause
+a significant performance impact given the relatively broad spans we're
+recommending (i.e. no high-throughput, per-event spans).
+
+One final concern when instrumenting with spans is the length of the span's
+lifetime. In early experiments, we ran into issues where extremely long-lived
+spans (e.g. covering the full runtime of a component) are opened and collect
+data continuously, never flushing because the span does not end. This lead to
+extreme amount of memory growth. The problem was greatly exaggerated by enabling
+`TRACE`-level logging, which resulted in an enormous number of log events
+collecting within the span. To address this, **we proposing pushing the
+outermost spans down from the lifetime of a component to an "iteration" of
+a component**. This will depend greatly on the specifics of each component, but
+the file source above can serve as an example. We also propose keeping these
+spans at the `INFO` level, where our logging is actually pretty quiet and there
+is far less potential for accumulating memory even if spans last longer than
+intended.
+
+In summary:
+
+- Spans should be component-focused, breaking down their operations by stages with at least two levels of hierarchy but up to three levels for components with more complex internals:
+    1. Single "iteration" of operation, on the order of seconds, as a baseline
+    2. High level stages of operation within an iteration, such as reading,
+       processing, sending
+    3. Interesting sub-stages of each operation, with a focus on units of work
+       with significant off-CPU time
```

Comment:
The view we have now from metrics and profiling is aggregate, in that it's done across multiple `Event`s flowing through vector. The value I would see from tracing, especially, would be something that's per-`Event`, so if we have an expensive path through vector we can actually see that path in practice. Is the proposal here more like our existing observation methods -- aggregate --  or do we get a more concrete view? That's where I'd see real value, being able to trace the flow of an `Event` through vector, especially considering the runtime impact.

Reply:
The tracing described here would not necessarily be per-`Event`, but I'm not sure that it'd really qualify as aggregate either. I'd also question the premise that per-`Event` is more "concrete" or valuable.

The goal is to provide more insight into how Vector is working internally, and Vector's internal operation is not necessarily focused on individual `Event`s. In the file source, for example, many of the issues we've fixed have occurred entirely before an `Event` is even constructed. In a transform, on the other hand, spans would likely cover the processing of a single event (or hopefully batch of events).

So the focus of this RFC is on tracing "operations", and it allows that to be defined however is most convenient for a given component. There's a bit of an implicit assumption that individual events are unlikely to be particularly interesting, at least relative to the degree by which some web requests can be more interesting than others, and that Vector has fewer, less variable paths of execution than your average web server. This could be taken as an argument that tracing is a less valuable tool for Vector.

Reply:
Makes sense.